File: Dynamic Programming/PPERM.cpp
// https://www.codechef.com/problems/PPERM

#include <bits/stdc++.h>
#define MAX 5000001
using namespace std;
typedef long long ll;

ll isprime[MAX];
ll primes[MAX];
ll perm[MAX];

void sievePerm(){
	fill(isprime, isprime+MAX, 1);
	primes[1]=0;
	perm[1]=1;

	isprime[1]=0;//one is not a prime number
	for(ll i=2; i*i<MAX; i++){
		if(isprime[i]){
			for(ll j=i*i; j<MAX; j+=i){
				isprime[j]=0; //making all multiples of i nonprime
			}
		}
	}
	ll MOD = 1000000007;
	for(ll i=2;i<MAX;i++){
		primes[i]=primes[i-1]+isprime[i];
		perm[i]=(perm[i-1]*(1+primes[i]))%MOD;
	}
}
int main()
{
    sievePerm();
	ll t;
	scanf("%lld",&t);
	while(t--)
	{
		ll n;
		scanf("%lld",&n);
		printf("%lld\n",perm[n]);
	}
	return 0;
}

File: Dynamic Programming/PPERM.py
# https://www.codechef.com/problems/PPERM

MAX=5000001
MOD=1000000007
def SieveOfEratosthenes(n):
    prime = [1 for i in range(n+1)]
    prime[1]=0
    p=2
    while(p*p <= n):
         
        if (prime[p] == 1):
            for i in range(p*2, n+1, p):
                prime[i] = 0
        p+=1
    return prime

def pperm():
    prime=SieveOfEratosthenes(MAX)
    perm=[1 for x in range(MAX)]
    primes=[0 for x in range(MAX)]

    for i in range(2,MAX):
        primes[i]=primes[i-1]+prime[i] # cumulating no. of primes
        perm[i]=(perm[i-1]*(primes[i]+1))%MOD
    
    return perm

perm=pperm()
t=int(input())
for _ in range(t):
    n=int(input())
    print(perm[n])
